azure/scope/machine.go

```diff
@@ -331,6 +331,22 @@ func (m *MachineScope) HasSystemAssignedIdentity() bool {
 	return m.AzureMachine.Spec.Identity == infrav1.VMIdentitySystemAssigned
 }
 
+// GetIdentityScope returns the specified identity scope, and generates a subscription scope if none is specified.
+func (m *MachineScope) GetIdentityScope() string {
+	if m.AzureMachine.Spec.Scope != "" {
+		return m.AzureMachine.Spec.Scope
+	}
+	return azure.GenerateSubscriptionScope(m.SubscriptionID())
+}
+
+// GetRoleDefinitionID returns the specified role definition ID, and generates a contributor role definition ID if none is specified.
+func (m *MachineScope) GetRoleDefinitionID() string {
+	if m.AzureMachine.Spec.RoleDefinitionID != "" {
+		return m.AzureMachine.Spec.RoleDefinitionID
+	}
+	return azure.GenerateContributorRoleDefinitionID(m.SubscriptionID())
```

Comment:
now that the field is exposed via the API this should get defaulted via the defaulting webhook so the default value is persisted and visible by users (same with scope)

Reply:
I'm having some trouble setting the default inside the webhook, since the scope and role definition ID requires the machine scope's subscription ID. I'm not sure if the AzureMachine webhook would have access to the subscription ID.

Reply:
Good point. The subscriptionID is part of the AzureCluster Spec, we'd have to fetch the owner AzureCluster object to be able to get to it. There is a way to do this by adding a Client to the webhook that can be used to fetch other objects in the k8s cluster (like https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/3f4543be33a1044a8fbb6746557baa4dd240e908/exp/api/v1beta1/azuremanagedmachinepool_webhook.go#L46) but it would require a bit more work to change the webhook for AzureMachine and AzureMachinePool since those aren't currently set up with a Client.

Reply:
After hacking away at this, I discovered that the machinepool is unable to retrieve the cluster name of the owner cluster when the default webhook is called. On this [line](https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/165a0d1653e3092b040c13f1939bfdb2534138c2#diff-435f85313e217cd2c7e0569bdace226dcb57ba2210dff964a2e511193010f809R291), the labels are empty. We were able to pin down the reason to the machine pool being declared and created directly from the spec, so the labels in [CAPI](https://github.com/willie-yao/cluster-api/blob/dff5c1a05a8aca1905b8a6bd8574f80f61f5f610/exp/internal/controllers/machinepool_controller_phases.go#L138) are not being set by the time the defaulting webhook is called. This is not the case for machines, as the machines are created by via a machine deployment and declared through a template.

Is there still a way to retrieve the clustername without using the labels field? Or is there a way to wait for the labels field to be populated. We were concerned about blocking or waiting in a webhook function though, so I'm not sure if there's a way around this. @CecileRobertMichon any thoughts?

Reply:
MachinePool.Spec has a required `clusterName` field, are you able to use that?

Reply:
I considered this option and I'm looking into how to get the MachinePool from the AzureMachinePool webhook. I don't have access to the machinepool scope or the name of the MachinePool from the webhook. I'm definitely just missing something though.

Reply:
I think you might be able to use GetOwnerMachinePool (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/92b41e932d17d416fc3c62831c8161337dd4d69c/controllers/helpers.go#L518) to get the MachinePool from the AzureMachinePool and then get the Cluster name from there